_posts/2024-02-05-cde-customization.adoc

```diff
@@ -0,0 +1,434 @@
+....
+
+### Install `ruby-lsp` Visual Studio Code extension
+
+Finally, the Ruby LSP extension, that failed to activate in STEP1, is now fully working:
+
+image::/assets/img/cde-customization/ruby-lsp-extension-fully-functional.png[]
+{nbsp} +
+Replacing Eclipse Che default image with a custom one allowed to include the right pre-requisites (ruby and rails) to the CDE. With that we are able to run the application and the VisualStudio Code ruby extension. 
+
+There are still a few properties of the CDE that can be improved (like increasing the CPU and memory, specifying the command to run the application, pre-fetch the dependencies etc...). We will cover those in the next session where we will use the Devfile for CDE more advanced configurations.
+
+[TIP]
+====
+In this part the image of the CDE has been replaced using a URL parameter. This can be done using a Devfile too as we will see in next session. An alternative to those approches is to change the image "on the fly" using `kubectl patch` against the DevWorkspace object:
+```bash
+DW_NAME="rails-sample" && \
+IMAGE="quay.io/mloriedo/rails-blog-cde:latest" && \
+kubectl patch dw "${DW_NAME}" --type='merge' \
+      -p '{"spec":
+            {"template":
+              {"components":
+                [
+                  { "name":"universal-developer-image", 
+                    "container":{"image":"'$IMAGE'"} }
+                ]
+              }
+            }
+          }' 
+```
+====
+
+## PART3: Change CDE properties with a Devfile [[part3]]
+
+In Part 2 we have seen how to customize a CDE with a new container image. That helps to include some development tools that are not in Eclipse Che default environment. But there are other properties of a CDE that a developer may want customize. Properties such as the memory or CPU requirements, containers to run in the CDE Pod (yes, a CDE can have multiple containers!), exposed endpoints, commands triggered at startup, at shutdown, etc...These properties can be customized using a Devfile and in this Part 3 we are reviewing how to do that. 
+
+### Define a Devfile for an Eclipse Che CDE
+
+A Devfile it's an https://www.cncf.io/projects/devfile/[open standard] to define Cloud Development Environments. It's a YAML file used to configure Eclipse Che CDEs.
+
+This is a simple Devfile that specifies the container image built in STEP 2 looks like:
+
+```yaml
+schemaVersion: 2.2.0
+metadata:
+  name: rails-sample
+components:
+  - name: devtools
+    container:
+      image: quay.io/mloriedo/rails-blog-cde:latest
+```
+
+It's reccomended to add the Devfile in the same git repository of the application developped (as https://github.com/l0rd/rails-sample/tree/devfile[the `devfile` branch of the `rails-sample` git repository]) using the name `.devfile.yaml`. When the Devfile is in the git repository it will evolve along with the application and it will be applied automatically whenever a developers provides the git repository URL to start a CDE.
```

Comment:
```suggestion
It's recommended to add the Devfile in the same git repository of the application developed (as https://github.com/l0rd/rails-sample/tree/devfile[the `devfile` branch of the `rails-sample` git repository]) using the name `.devfile.yaml`. When the Devfile is in the git repository it will evolve along with the application and it will be applied automatically whenever a developer provides the git repository URL to start a CDE.
```